enhancements/multi-arch/multiarch-manager-operator.md

```diff
@@ -0,0 +1,773 @@
+    MutatingAdmissionWebhook ->> APIServer: patched pod
+    deactivate APIServer
+
+    par
+        APIServer -->> Scheduler: Notify pod event
+        alt len(schedulingGates) > 0
+        Scheduler -->> Scheduler: hold in the queue until no scheduling gates are set
+        end
+
+    and 
+        APIServer -->> PodPlacementController: Notify pod event
+
+        alt schedulingGatePresent
+            activate PodPlacementController
+            PodPlacementController ->> Image Registry: Inspect Container Images
+
+            Image Registry -->> PodPlacementController: supported architectures
+            PodPlacementController ->> PodPlacementController: Intersect Supported Architectures
+            PodPlacementController ->> Pod: PatchNodeAffinity
+            Note over PodPlacementController: inspection errors are ignored<br>to allow the removal of the scheduling<br>gate.
+            deactivate PodPlacementController
+            PodPlacementController ->> Pod: RemoveSchedulingGate
+        end
+    end
+    Note over Scheduler: no more schedulingGates
+    Scheduler ->> Scheduler: move Pod to activeQ <br> and other plugins
+
+```
+#### Deletion of the CR
+1. The cluster admin deletes the CR
+2. All resources related to the operator - the controllers, webhook service and
+   pods automatically get deleted
+3. Any pods stuck in the scheduling gated phase will have to be patched manually
+   to resume scheduling (by removing the scheduling gate)
+4. Recreating the CR again will install all related resources
+
+#### Operator uninstallation
+- Operator uninstallation removes the operator. If the option to uninstall
+  operands is selected, all resources are removed. 
+
+#### Variation
+- *Pods ignored during scheduling*: the `PodPlacementConfig` CR allows a
+  namespace exclusion list. Any pods belonging to the excluded namespaces will
+  be gated at admission time and will be ignored by the `PodPlacementController`
+- *NodeAffinity block already set in the gated pod*: when patching the node
+  affinity, any node selector term having a `matchExpression` element using the
+  key `kubernetes.io/arch` is preserved. For all the other node
+  selector terms, we append the architecture-specific `matchExpression`.
+- *Image inspection failure*: in the node affinity patching workflow, if there
+  are any errors during the image inspection process (authentication issues,
+  network errors, ...), the scheduling gate is removed, and no changes are made
+  to the pod. In this case, the scheduling will only be architecture-aware if
+  the user submitting the pod defines the affinity predicates explicitly.
+
+### API Extensions
+
+#### Mutating Admission Webhook and Scheduling Gate 
+The multiarch-manager-operator reconciles not-yet-scheduled pods using the
+scheduling readiness[^1] and pod mutable scheduling directives[^2] features.
+Since Kubernetes 1.27[^3] promoted those features to beta, no additional changes
```

Comment:
the [pod scheduling readiness](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3521-pod-scheduling-readiness) and[ pod mutable scheduling directives](https://github.com/kubernetes/enhancements/blob/ac65b5853a9a5b8fb77a8c3069c07780057a4c9e/keps/sig-scheduling/3838-pod-mutable-scheduling-directives/README.md) KEPs were implemented and graduated to beta in 1.27, so the required feature is already in.

Reply:
Is there no precedence in Openshift where we have GA'ed a feature with a beta API? Is that a possibility? Looking at https://kubernetes.io/blog/2020/08/21/moving-forward-from-beta/, it seems there is a possibility that a feature could be stuck in beta.. we are going to work upstream with the sig-scheduling folks to understand what it would take to GA this API.

Reply:
> Is there no precedence in Openshift where we have GA'ed a feature with a beta API?

Just because we've made a mistake in the past does not mean we must make a mistake in the future.  Determining which beta features we're willing to depend upon requires consideration of the likelihood of progression, whether we are investing in promoting that feature, and other upstream consideration.  This particular one I'm not familiar with and I suggest you contact @soltysh about discussing on the workloads arch call.

Reply:
yes i'm not saying we should include this indiscriminately. I'm just worried about the feature getting stuck in beta.